### PR TITLE
Refactor template tests CI pipeline

### LINKED ISSUE #1025: Update test-templates pipeline with more input vars
I would like to be able to configure all the inputs to the test-templates.sh script via environment variables.

I would like to be able to tell the pipeline to use the azd build from the target PR instead of building it from main.

### PR DESCRIPTION
**Pull request by weikanglim**
This change improves various aspects of the template test pipeline to make it reusable in multiple testing scenarios.

There are two set of changes:
1. Template test pipeline is now refactored as a job template, in preparation for reuse across other pipelines.
2. Template test pipeline is now fully parameterized that allows tests to target: a specific azd version, specific templates, and specific template branch.

With the two changes above, we can start integrating template testing into different situations such as:
- Test a new `azd` PR build against templates
- Chain the template test jobs into our PR validation for template related changes. In fact, we can filter out only templates that have changed given the flexible parameterization above.

Fixes #1025

### CODE REVIEW

**Comment by weikanglim** [comment]
Successful run with all templates: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1971325&view=results
Successful run with filtered templates: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1971354&view=results

File: eng/pipelines/templates/jobs/run-template-tests.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    strategy:
+      matrix: $[ dependencies.TemplateTest_CreateMatrix.outputs['Generate.Matrix'] ]
+      maxParallel: 10
+
+    pool:
+      name: azsdk-pool-mms-ubuntu-2004-general
+      vmImage: MMSUbuntu20.04
+
+    #container: $[ variables['AzdContainerImage'] ]
```
Comment (ellismg) [inline eng/pipelines/templates/jobs/run-template-tests.yml:86]: Can we delete?

**Review by ellismg — APPROVED** [review]
I love it!

File: eng/pipelines/templates/jobs/run-template-tests.yml (hunk 1 above)
Comment (weikanglim) [reply eng/pipelines/templates/jobs/run-template-tests.yml]: Created #1116 to track

**Comment by weikanglim** [comment]
I ran `/azp run azure-dev - template - tests` which failed (as expected) since there were no CLI changes, and that version of `azd` wasn't published. The failing test is expected. Will need to override since `template-tests` are no included as part of check-enforcer.

**Pull request merged**